### ISSUE TITLE
Traffic speed doesnt add up

### ISSUE THREAD

**Issue by benisai** [mstojek/nlbw2collectd#9]
Hi, I know your LUA file only setups a method to collect the data nlbw creates, but I cant seem to get my data to match what my speed test shows. 

I've been trying to find a replacement for iptmon and found your repo. The only change I made was setting the interval to 10 seconds on both the lua.conf and nlbwmon config. I ran a speed test and got about 17 down and up. but the graph only shows 600 bytes

<img width="766" alt="image" src="https://github.com/user-attachments/assets/87b5ddb4-fcc0-4cc3-97ac-4b4517bec7db" />
<img width="976" alt="image" src="https://github.com/user-attachments/assets/c63f6869-683d-44d1-ba85-3f43cb8e4970" />
<img width="982" alt="image" src="https://github.com/user-attachments/assets/9e1cc93f-6a28-4f1a-9314-98303492b8c3" />

**Comment by mstojek** [comment]
My script is pretty simple - it just executes command `/usr/sbin/nlbw -c json -g ip` parses the output and sends data to collectd. 
We need to debug what is happening here. First question is - are you using IPv6? My script is not collecting data for IPv6 - it can be added if needed. I do not use IPv6 so I did not add it.
Second let us do quick debug:
1. Restart collectd with `/etc/init.d/collectd restart` command
2. Execute command `logread | grep collectd` from the command line, do we see any collectd errors after collectd reset? 
3. Execute command `/usr/sbin/nlbw -c show -g ip` and note what are the current traffic counters for your hosts
4. Start speed test
5. Execute command `/usr/sbin/nlbw -c show -g ip` and note what are the traffic counters for your hosts - do we see increase here?
6. Execute command `logread | grep collectd` from the command line, do we see any collectd errors after the test?

**Comment by benisai** [comment]
I need some more details on how you get the _rate_

I looked over the code again, nlbw -c show -g ip will output the **total** gathered down and up, not the rate per second. How are you gathering the rate from nlbw if its not offered as an output?



For example 871443624 is 831.07 MB (in binary) , and this is what you use local tx_bytes_modulo = tx_bytes % 2147483647. 

<img width="961" alt="Image" src="https://github.com/user-attachments/assets/ebb022a0-07d8-4978-b02c-db8abaae4617" />


```
root@GL-AXT1800:~# nlbw -c show -g ip
             IP      Conn.   > Downld. ( > Pkts. )      Upload (   Pkts. )
     10.0.3.245       261    831.07 MB ( 726.43 K)   127.80 MB ( 174.29 K)
     10.0.3.244     1.63 K   240.35 MB ( 113.02 K)    14.51 MB (  57.96 K)
     10.0.3.174       494     97.50 MB (  55.06 K)     5.73 MB (  19.17 K)
  192.168.64.74     1.13 K     4.25 MB (   4.19 K)   309.22 KB (   4.11 K)
       10.0.3.1         4          0 B (       0 )     2.73 KB (      21 )
   10.24.27.135         1          0 B (       0 )       552 B (       4 )
     10.0.3.240         1          0 B (       0 )     1.30 KB (       6 )
root@GL-AXT1800:~# 
root@GL-AXT1800:~# 
root@GL-AXT1800:~# 
root@GL-AXT1800:~# nlbw -c json -g ip
{"columns":["ip","conns","rx_bytes","rx_pkts","tx_bytes","tx_pkts"],"data":[["10.0.3.245",261,871443624,743874,134009318,178479],["10.0.3.244",1681,252079922,115767,15222383,59373],["10.0.3.174",494,102239532,56383,6012373,19635],["192.168.64.74",1167,4457602,4293,316650,4218],["10.0.3.1",4,0,0,2799,21],["10.24.27.135",1,0,0,552,4],["10.0.3.240",1,0,0,1332,6]]}root@GL-AXT1800:~# 
```

**Comment by benisai** [comment]
I'm very confident this project does not accomplish what you want. There is no way to get the rate from nlbw, at least not by the method you're using. If I'm wrong, please let me know. 

Looking at the nlbwmon repo code, we see that running the command will dump data from the database not the rate. 

![Image](https://github.com/user-attachments/assets/c7985462-f531-4e52-938d-504e3abba1da)

**Comment by benisai** [comment]
if you want to get the rate, like iptmon, you need to setup chains. nft-qos does this and can get the rate. check out the scripts. 
https://github.com/rosywrt/nft-qos/blob/master/package/files/lib/core.sh
https://github.com/rosywrt/nft-qos/blob/master/package/files/lib/monitor.sh

**Comment by benisai** [comment]
I dont know if this will work, but I passed those scripts in CGPT and it gave me this

> Dynamic IP Fetching:
> 
> The script uses /tmp/dhcp.leases to fetch the list of connected IP devices. You can switch to parsing /proc/net/arp if you prefer the ARP table.
> Each connected IP is added to the nftables monitoring setup dynamically.
> Monitoring All Devices:
> 
> For each connected device, download and upload rates are calculated and displayed.
> Counters Stored in Memory:
> 
> The script uses associative arrays (prev_download and prev_upload) to track the previous counters for each IP dynamically.
> Example Usage
> Save the script to a file, e.g., dynamic_nft_monitor.sh.
> Make it executable:
> chmod +x dynamic_nft_monitor.sh
> Run the script:
> ./dynamic_nft_monitor.sh
> Notes
> Ensure the nftables package is installed and working on your OpenWRT router.
> Adjust the path to dhcp.leases or use an alternative method (/proc/net/arp) to fetch device IPs if needed.
> This script continuously monitors traffic rates for all connected devices.
> 




```
#!/bin/sh
#
# Dynamic Script to Monitor Traffic Rates for All Devices on OpenWRT
#

# Define constants
NFT_TABLE="nft-rate-monitor"
NFT_CHAIN_DOWNLOAD="download"
NFT_CHAIN_UPLOAD="upload"

# Initialize nftables setup (creates table and chains if they don't exist)
init_nft_rate_monitor() {
    nft add table inet "$NFT_TABLE" 2>/dev/null || true
    nft add chain inet "$NFT_TABLE" "$NFT_CHAIN_DOWNLOAD" { type filter hook postrouting priority 0\; } 2>/dev/null || true
    nft add chain inet "$NFT_TABLE" "$NFT_CHAIN_UPLOAD" { type filter hook input priority 0\; } 2>/dev/null || true
}

# Add a rule to monitor rates for an IP address
add_monitor_rule() {
    local ip="$1"

    # Add rules for download and upload counters
    nft add rule inet "$NFT_TABLE" "$NFT_CHAIN_DOWNLOAD" ip daddr "$ip" counter 2>/dev/null || true
    nft add rule inet "$NFT_TABLE" "$NFT_CHAIN_UPLOAD" ip saddr "$ip" counter 2>/dev/null || true
}

# Retrieve and print counter rates for an IP address
get_rate() {
    local ip="$1"
    local chain="$2"

    # Retrieve the counter data
    nft list chain inet "$NFT_TABLE" "$chain" | grep "$ip" | awk '{print $NF}'
}

# Monitor traffic rates for all connected devices
monitor_all_traffic() {
    local connected_ips
    local prev_counters
    declare -A prev_download
    declare -A prev_upload

    while true; do
        # Get list of connected devices' IPs (via DHCP leases or ARP table)
        connected_ips=$(cat /tmp/dhcp.leases | awk '{print $3}') # Adjust path if needed

        for ip in $connected_ips; do
            # Add rules for any new IPs
            add_monitor_rule "$ip"

            # Get current counters
            curr_download=$(get_rate "$ip" "$NFT_CHAIN_DOWNLOAD")
            curr_upload=$(get_rate "$ip" "$NFT_CHAIN_UPLOAD")

            # Default values if counters are not yet available
            curr_download=${curr_download:-0}
            curr_upload=${curr_upload:-0}

            # Calculate rate (bytes per second)
            rate_download=$((curr_download - ${prev_download[$ip]:-0}))
            rate_upload=$((curr_upload - ${prev_upload[$ip]:-0}))

            # Update previous counters
            prev_download["$ip"]=$curr_download
            prev_upload["$ip"]=$curr_upload

            # Print the rates
            echo "IP: $ip | Download Rate: ${rate_download} B/s | Upload Rate: ${rate_upload} B/s"
        done

        # Wait for 1 second before the next check
        sleep 1
    done
}

# Main script logic
main() {
    # Initialize nftables
    init_nft_rate_monitor

    # Start monitoring traffic rates for all devices
    monitor_all_traffic
}

# Run the main function
main

```

**Comment by mstojek** [comment]
My script is very easy: It dumps current state of the nlbwmon traffic counters and sends them to collectd. This is Collectd (or rather RRD Tool) that calculates the rate, total traffic data, downsamples them for long storage. If you want to know how the traffic counter is used to calculate rate take a look at [RRD Tool](https://oss.oetiker.ch/rrdtool/tut/rrdtutorial.en.html) documentation for details. 

The line that you found `tx_bytes_modulo = tx_bytes % 2147483647` is a workaround for issue that on some systems interface for integer number between lua and collectd is only 32 bit. So we can not send to collectd counter bigger than ~2GB. 
For Openwrt 21.0 I prepared [patch](https://github.com/mstojek/nlbw2collectd/blob/main/collectd-lua-patch/999-lua_use_numbers_instead_integers_for_counters.patch) to solve this issue. Applying this patch requires Openwrt compilation. When using this patch you can remove the workaround and change this line to: `tx_bytes_modulo = tx_bytes`

To summarize: My script does what it is intended to do. It allows you to create collectd graphs showing rate and total traffic usage. By exporting it to Grafan and InfluxDB you can get pretty detailed statistics (comparing to Collectd). The workaround for 32bit issue that I use causes unfortunately some innacuracy with traffic data, not an issue if the traffic is at level of 10Mb/s, but for 1Gb/s you can get seriously wrong data. E.g. for 100Mb/s constant traffic you will get error being in worst case about 17%. Of course on average it will be much less (about 3% I think). But for gigabit constant traffic the counter could be wrong by more than 100% - in this case you can not use the workaround:  `tx_bytes_modulo = tx_bytes % 2147483647`.

**Comment by benisai** [comment]
Explain the collectd rate conversion to me. 

I don't see how collectd can get the rate if you're exporting the TOTAL usage from the database.  (Unless I've collected is doing some type of difference; or before and after math). 

And when you say current counters, where are you getting this info from? Because nlbw only stores totals in the database. 

I read a bit on rrd but I would like you to explain the thought process on the project (script with collectd) rate part so we can get a better understanding.

**Comment by mstojek** [comment]
This is collectd plugin. Collectd calls my plugin every 30 seconds. So every 30 seconds collectd gets current traffic counters. By calculating difference between current counters and previous period counters collectd calculates rate.

**Comment by benisai** [comment]
Awesome. Thats all I needed to know :)
